pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java

```diff
@@ -0,0 +1,134 @@
+   * -columnStats, which stores column information, collected from kafka or staging pinot table.
+   * ```
+   * For columns stats, we need to gather the following stats
+   * ```
+   * -cardinality, a required information unique combination of primary keys.
+   * -primaryKeySize, it uses for calculating BytesPerKey
+   * -comparisonColSize, it uses for calculating BytesPerValue
+   * ```
+   */
+  @POST
+  @Path("/heapUsage")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Estimate memory usage for an upsert table", notes =
+      "This API returns the estimated heap usage based on primary key column stats."
+          + " This allows us to estimate table size before onboarding.")
+  public String estimateHeapUsage(@QueryParam("tableConfig") String tableConfigStr,
+      @QueryParam("schema") String tableSchemaStr, @QueryParam("columnStats") String columnStatsStr) {
+    TableConfig tableConfig;
+    Schema schema;
+    ColumnStats columnStats;
+    try {
+      columnStats = JsonUtils.stringToObject(columnStatsStr, ColumnStats.class);
+    } catch (IOException e) {
+      String msg = String.format("Invalid column stats json string: %s", columnStatsStr);
+      throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST, e);
+    }
+
+    try {
+      tableConfig = JsonUtils.stringToObject(tableConfigStr, TableConfig.class);
+    } catch (IOException e) {
+      String msg = String.format("Invalid table config json string: %s", tableConfigStr);
+      throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST, e);
+    }
+    try {
+      schema = JsonUtils.stringToObject(tableSchemaStr, Schema.class);
+    } catch (IOException e) {
+      String msg = String.format("Invalid table schema json string: %s", tableSchemaStr);
+      throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST, e);
+    }
+
+    int bytesPerKey = 0;
+    ObjectNode resultData = JsonUtils.newObjectNode();
+    List<String> primaryKeys = schema.getPrimaryKeyColumns();
+    // Estimated key space, it contains primary key columns
+    for (String primaryKey : primaryKeys) {
+      FieldSpec.DataType dt = schema.getFieldSpecFor(primaryKey).getDataType();
+      if (dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST || dt == FieldSpec.DataType.MAP) {
+        String msg = "Not support data types for primary key columns";
+        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+      } else if (dt == FieldSpec.DataType.STRING) {
+        bytesPerKey += columnStats.getPrimaryKeySize();
+      } else {
+        bytesPerKey += dt.size();
+      }
+    }
+    // Estimated value space, it contains <segmentName, DocId, ComparisonValue(timestamp)>
+    int bytesPerValue =
+        tableConfig.getUpsertConfig().getComparisonColumn() != null ? 52 + columnStats.getComparisonColSize() : 64;
```

Comment:
If comparison column is fixed width, no need to provide it in `columnStats`

Reply:
gotcha, thanks Jackie. I will infer the comparison size from Datatype. 
`    String comparisonColumn = tableConfig.getUpsertConfig().getComparisonColumn();
    if (comparisonColumn != null) {
      FieldSpec.DataType dt = schema.getFieldSpecFor(comparisonColumn).getDataType();
      if (dt == FieldSpec.DataType.STRING || dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST
          || dt == FieldSpec.DataType.MAP) {
        String msg = "Not support data types for the comparison column";
        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
      } else {
        bytesPerValue = 52 + columnStats.getComparisonColSize();
      }
    }`